src/WebJobs.Script/Description/FunctionGenerator.cs

```diff
@@ -110,31 +119,47 @@ public static Type Generate(string functionAssemblyName, string typeName, Collec
                 il.Emit(OpCodes.Ldloc_1);
                 il.Emit(OpCodes.Ldloc_0);
                 il.Emit(OpCodes.Callvirt, invokeMethod);
-
-                // Copy back out and ref parameters
-                for (int i = 0; i < function.Parameters.Count; i++)
+                
+                
+                if (function.Parameters.Any(p => p.Type.IsByRef))
                 {
-                    var param = function.Parameters[i];
-                    if (!param.Type.IsByRef)
+                    // Store the result into a local Task
+                    il.Emit(OpCodes.Stloc_2);
+
+                    // Load the result Task
+                    il.Emit(OpCodes.Ldloc_2);
+
+                    // We need to wait on the function's task if we have any out/ref
+                    // parameters to ensure they have been populated before we copy them back
+                    il.Emit(OpCodes.Callvirt, typeof(Task).GetMethod("Wait", Type.EmptyTypes));
```

Comment:
We need to wait here, as this method can't complete asynchronously, otherwise the out/refs won't be correctly set. I guess I can configure the task awaiter on the original call so that it doesn't capture the synchronization context. Thoughts?